### ISSUE TITLE
Dev toolbar ideas

### ISSUE THREAD

**Issue by timgl** [PostHog/posthog#741]
Let's use wisdom of the crowds: how can we surface a lot of the information that's inside PostHog within people's websites/apps.

The kind of information that's relevant
- Pageviews/unique users
- Sessions
- Dropoffs
- What users do after this page (paths)
- What users generally do on this page (Heatmaps)
- How any of the above has changed over time

The way to display this I've left intentionally vague. Let's come up with some cool new ways of showing/accessing all this information

Ideas could be sketches, wireframes, mockups or just descriptions. Let's have some ideas by next Monday (18 may)

**Comment by EDsCODE** [comment]
**Key Qualities:**
- The need for a developer to use tool to see stats is a once per session type behavior as opposed to consistently needing it to be visible and updated throughout developing session. This means the toolbar must be easy to get completely out of the way after using.
- The toolbar must feel more worth it than pulling up the dashboard. Interactivity of toolbar is key here. Allowing users to be able to actually see how the numbers are related to the elements on the page

**Initial brainstorm:** (Similar to inspector toolbar for people to add actions currently)
Create a browser extension that functions as a toolbar used during development. The extension will be clickable when it detects you're on one of your whitelisted websites and by clicking it you will pull up a side menu that works as an inspector/HUD for the site elements. 

 (TBD if possible with an extension) The heatmap tab will gray out the browser and colorize the elements that are high usage to visualize the heavily used elements. More of the phase 2 features could be added as tabs depending on limitations of a browser extension.

Rough sketch:
<img width="1083" alt="Screen Shot 2020-05-17 at 7 50 50 PM" src="https://user-images.githubusercontent.com/13127476/82163310-5f43d580-9878-11ea-99e2-505ea302cfcf.png">

Netflix party has a similar toolbar-esque concept
<img width="732" alt="Screen Shot 2020-05-17 at 7 53 33 PM" src="https://user-images.githubusercontent.com/13127476/82163309-5f43d580-9878-11ea-955b-9136e1202e63.png">

**Comment by mariusandra** [comment]
Hi! Some thoughts on this in no specific order:

* What is or should there be a difference between the "dev toolbar" and the current "select element on site" toolbar/modal/popup? I think we don't have the resources to make two really good toolbars, so it should be one and the same. Meaning that when I have the "dev toolbar" open, I can also add new actions, etc.
* Then information that I'm interested in is to have a visual overview of what did most people do who opened this page, plus the ability to set the date filter when I'm actively running experiments (e.g. I did changes 2 days ago and I want to see the actions since then).
* I imagine that we gather all events where $current_url = window.location.href and then aggregate all the HTML elements/selectors that were most clicked/interacted with. 
* I'm not sure if heatmaps are the best way to display this information, since traditionally in heatmaps each "dot" represents one interaction. However we don't track mouse click coordinates, so what if there's a 800px x 100px banner that gets 50% of the clicks and a 100px x 20px link that gets the other 50%. How to show that with a heatmap?
* I'd much rather prefer a "(1)", "(2)" style overlay mode. One cool effect is to zoom out the page in this mode and show the overlaid information next to it.

Here's a really rough draft of this concept:

![image](https://user-images.githubusercontent.com/53387/82180626-34419c00-98e1-11ea-86fc-23650b0f7345.png)

This is just one view for the toolbar. There could be tabs or other links to toggle between 1) what people click, 2) what page they go to next, 3) list of all custom actions on this page, 4) list of all events that people do that aren't custom actions, 5) bunch of graphs for various things

For the "dev" part of this "dev toolbar", I imagine I could run the same code in localhost. It would overlay as well information that people do in production mode, plus I would have a chance to add custom actions. I imagine I would add actions for all the new things/buttons I add to the page, just to track their performance over time.

E.g. I'm adding a new CTA button in localhost, I tag clicks on it directly as "CTA experiment nr 2" and then once the site is live, this data is automatically available for me.

In any case, this is a very rough brainfart to get the ball rolling.

**Comment by fuziontech** [comment]
I have always been a fan of a heatmap if you are going to overlay anything on the user's website.
Something like the classic crazyegg heatmap:
https://www.crazyegg.com/heatmap

A few spins on this could be showing an inverse heatmap where you show cooler (bluish) colors for where user go before dropping off a funnel. (would be funny if that turned out to be terms page or the about page for a company)

Showing and segmenting a heatmap view by how far users got into the funnel or cohort would be interesting as well. This would especially be interesting if you were comparing cohorts of users within your app to see how a veteran user uses an app vs a new user. Doing a side by side of cohorts would be potentially interesting as well.

I think the really important thing is keeping the view simple and intuitive. Only show things that are made better by showing the context of the page itself. Generally this will boil down to someone trying to improve the UX of the site so keeping that audience in mind is key. I think this would become super relevant if a user was using Posthog along with A/B testing via tags on events. You could segment and do a side by side heatmap of A variant next to B variant and see intuitively how the experiment affects user interaction.

**Comment by jamesefhawkins** [comment]
People will look at individual pages next to data to try to work out "why" users are not converting.

To be really useful, we need to give the answer to that question. For example, I love [Google PageSpeed](https://developers.google.com/speed/pagespeed/insights/?url=posthog.com&tab=desktop) as it just tells me what I need to know.

The answer will be from the following. This is *really* exciting since PostHog will often have lots of personal information contained within it, to give better answers than Google Analytics could, for example:

* The page content actually helps the user so they leave, happy.
  * Shown by: stickiness, session times
* Users navigate themselves away from the goal you have in mind
  * Shown by heatmaps like Marius' suggestion above
* The page doesn't cater to what a user needs
  * Shown by: a high bounce rate, so that should go onto the toolbar. Often the bounce rate may depend on the following factors:
    * The page doesn't display properly
      * Shown by: being able to segment the bounce rate by property (device) or very short session times
    * The page doesn't meet expectations with where the user came from
      * Shown by: being able to segment the bounce rate by property (UTM tags or referrer)
    * The page is confusing
      * Shown by: *no other obvious factors* - often a lack of CTA or poor messaging
    * The page isn't compelling enough for the user to proceed
      * Shown by same as above
    * The user isn't a good fit
      * Shown by: being able to segment the bounced users by property (whatever user fields are available - perhaps some profile data from the user, and if that doesn't exist - alerting the PostHog user that they should send this to PostHog if available to make this possible)

It would be very cool to start suggesting the above ie "users on mobile have a much higher bounce rate on this page".

## Getting the data right

Products and pages change, a lot. Especially with startups.

1. This makes it more important to show graphs where possible
1. Users need to be able to filter by date range
  1. A really powerful extension would be to integrate with GitHub to visualize all the PRs over time that changed the product


## The discussion above

The ideas from above that are strongest IMO:

* Interactivity - it should feel like it's working with the website/application not just displaying data that you could get in the app
* need to be able to close it nicely (specifically, it needs NOT to affect CSS when closed for UX people)
  * Random thought: if it could feel more like the native inspector in the browser rather than say the wordpress dropdown, it'd be a lot nicer to work (I often would use incognito when developing on wordpress as i'd be worried the toolbar would affect the CSS). Although that'd involve building this as an extension, which probably wouldn't be great since it'd make the task harder than necessary to start with.
* it must work on localhost since that's where people develop
* I like Marius' suggestion of 'numbers' not pixel-perfect heatmaps. I don't think they'd add much value
* user / cohort side-by-sides / feature flag side-by-sides are a very cool idea and something we could do quite uniquely

**Comment by timgl** [comment]
Love everyones suggestions

- The secret about those heatmaps is they don't actually use x/y coordinates but elements (otherwise they'd have to marry up all sorts of different screen sizes). I suggest we also add heatmaps, purely so we can say 'we have heatmaps'. Functionally I much prefer Marius' version.

This is the low quality mock up I've come up with, based largly on Marius' work
![image](https://user-images.githubusercontent.com/1727427/82442303-4e24df80-9a97-11ea-9804-777cdb9886a2.png)

This multi tab approach means we'll be able to stuff quite a lot of functionality and data in without overwhelming the user.

**Comment by EDsCODE** [comment]
Here's an initial draft of what I got going to tidy up what @timgl started. I also added the "inspector" thing as an idea for greater interactivity. The description should give an idea of what I mean. Will continue hashing out (need more specifics on date/editing etc.)

<img width="388" alt="Screen Shot 2020-05-20 at 9 48 00 PM" src="https://user-images.githubusercontent.com/13127476/82514615-332c8c80-9ae4-11ea-901b-1f1cd8bdee87.png">

**Comment by timgl** [comment]
I think this is great. The inspector doesn't make too much sense to me outside the "Actions" tab, as all the elements will already be highlighted on the page and will be in the rows below it.

In terms of the other 2 tabs, "Actions" should function very much like the current toolbar, in that it'll allow you to create/edit actions. Dashboards I think should just be a list of all of the users' dashboards in small mode in easy reach.

**Comment by EDsCODE** [comment]
Ohh, I see what you're going for now. So the stats (heatmap) and actions page should be focused on interacting with the page at hand and the dashboard page will be an overview/quick look page?

**Comment by timgl** [comment]
Yep exaclty!

**Comment by fuziontech** [comment]
👍 I think this makes a lot of sense. Using the elements is perfect for the heatmap.

Having a bounce rate metric on the panel would be nice too.

**Comment by mariusandra** [comment]
Hi all! I tried to implement code that would produce a result resembling this:

![](https://user-images.githubusercontent.com/1727427/82442303-4e24df80-9a97-11ea-9804-777cdb9886a2.png)

I got something working, but it's far from ideal. It might be good enough and we can improve on this once there are better ways that we come up with. Let me know what you think!

All the code blocks below should just be copied to the devtools console if you want to try them out.

## Only 100% working solution

The only approach that works 100% of the time is to have a floating widget on the page, like we have now:

![image](https://user-images.githubusercontent.com/53387/82883379-a00bb080-9f42-11ea-9835-00e04d728e62.png)


## Gallery of failed attempts

The following methods yielded unusable results:

### Reloading the page in an iframe

I tried a simple approach where we nuke all the content on the screen and load the page again in an iframe, possibly keeping the page URL in sync between the iframe and the content.

Turns out that's a bad idea since many websites try to break out of frames for security and aesthetic reasons:

```javascript
document.write('');
document.close();

var iframe = document.createElement('iframe');
iframe.src = location.href
document.body.appendChild(iframe);
```

![screencast 2020-05-26 11-15-33](https://user-images.githubusercontent.com/53387/82883186-633fb980-9f42-11ea-9bfb-148091e639d2.gif)

So this is a no-go. Overriding/disabling these kinds of frame escapes would break too many things and it's not worth it, if even possible.

### Setting the `<body>` width and putting the toolbar on the right

Just setting the width of the body (and/or html) tags yields bad results, as the page might have absolutely positioned elements:

![image](https://user-images.githubusercontent.com/53387/82884083-8ae35180-9f43-11ea-90a3-ec9ffac18e2f.png)

... or rely on CSS media queries manually setting "width" of elements based on the viewport. The simplest is a rule like "if screen width > 1280, div width = 1280":

![image](https://user-images.githubusercontent.com/53387/82883944-66877500-9f43-11ea-9857-56f5abe686b6.png)

### Making a new [div id='fakebody'] and moving <body> inside it.

There are too many issues with this approach unfortunately.

## What works, kind of.

Thus the only way to reliably "zoom out" the page was to actually zoom it out. For example by setting:

```javascript
const zoom = 0.7
window.document.body.style.transform = `scale(${zoom})`
window.document.body.style.transformOrigin = `top left`
```

![image](https://user-images.githubusercontent.com/53387/82884482-11982e80-9f44-11ea-8728-3a4807cc441b.png)

However when I wanted the "picture in picture" effect by moving the "body" 30px down and to the right, I ran into tiny problems. Using the following CSS it mostly worked:

```javascript
const zoom = 0.7
window.document.querySelector('html').style.background = '#a1a3ae'
window.document.body.style.transform = `scale(${zoom}) translate(${20/zoom}px, ${20/zoom}px)`
window.document.body.style.transformOrigin = `top left`
window.document.body.style.width = `100vw`
window.document.body.style.minHeight = `100%`
if (!window.document.body.style.backgroundColor) {
  window.document.body.style.backgroundColor = 'white'
}
```

![image](https://user-images.githubusercontent.com/53387/82885194-18737100-9f45-11ea-9597-5b3ced406b91.png)

However apparently setting `transform` on a property [nukes](https://stackoverflow.com/a/28157774) `position: fixed` attributes. Thus it killed all floating sidebars/headers on all sites.

![screencast 2020-05-26 11-45-28](https://user-images.githubusercontent.com/53387/82886213-7fddf080-9f46-11ea-86c8-f1def95fbc61.gif)

Which is a shame, as I was planning to use `position: fixed` to position the toolbar itself on the right. To get around that, I had to create a window.scroll listener, that looks for all elements with position: fixed and then adds `marginTop` to them. 

The result is not beautiful, but it seems to do the job *most* of the time:

```javascript
const zoom = 0.7
window.document.querySelector('html').style.background = '#a1a3ae'
window.document.body.style.transform = `scale(${zoom}) translate(${20/zoom}px, ${20/zoom}px)`
window.document.body.style.transformOrigin = `top left`
window.document.body.style.width = `100vw`
window.document.body.style.minHeight = `100%`
if (!window.document.body.style.backgroundColor) {
  window.document.body.style.backgroundColor = 'white'
}
window.addEventListener('scroll', function(e) {
  [...document.body.getElementsByTagName("*")].filter(
    x => getComputedStyle(x, null).getPropertyValue("position") === "fixed"
  ).forEach(e => {
    const h = (window.pageYOffset + (window.pageYOffset > 20 ? -20 : 0)) / zoom
    e.style.marginTop = `${h}px`
    e.setAttribute('data-posthog-fix-fixed', 'yes')
  });
  [...document.querySelectorAll("[data-posthog-fix-fixed=yes]")].filter(
    x => getComputedStyle(x, null).getPropertyValue("position") !== "fixed"
  ).forEach(e => {
    e.style.marginTop = 0
  })
})
```

![screencast 2020-05-26 11-51-00](https://user-images.githubusercontent.com/53387/82886929-8a4cba00-9f47-11ea-96c1-fb9640b1a2db.gif)

The bottom `fixed` div is still broken, but I'm sure it can be fixed with some additional hack.

Then to add a toolbar, I created a shadow root and added a random div. Similar to how the current toolbar does it.

This is the final result:

```javascript
const zoom = 0.7
window.document.querySelector('html').style.background = '#a1a3ae'
window.document.body.style.transform = `scale(${zoom}) translate(${20/zoom}px, ${20/zoom}px)`
window.document.body.style.transformOrigin = `top left`
window.document.body.style.width = `100vw`
window.document.body.style.minHeight = `100%`
if (!window.document.body.style.backgroundColor) {
  window.document.body.style.backgroundColor = 'white'
}
window.addEventListener('scroll', function(e) {
  [...document.body.getElementsByTagName("*")].filter(
    x => getComputedStyle(x, null).getPropertyValue("position") === "fixed"
  ).forEach(e => {
    const h = (window.pageYOffset + (window.pageYOffset > 20 ? -20 : 0)) / zoom
    e.style.marginTop = `${h}px`
    e.setAttribute('data-posthog-fix-fixed', 'yes')
  });

  [...document.querySelectorAll("[data-posthog-fix-fixed=yes]")].filter(
    x => getComputedStyle(x, null).getPropertyValue("position") !== "fixed"
  ).forEach(e => {
    e.style.marginTop = 0
  })
})

customElements.define('posthog-toolbar', class extends HTMLElement {
  connectedCallback() {
    const shadow = this.attachShadow({mode: 'open'});
    const style = document.createElement('style');
    shadow.appendChild(style);
    shadow.innerHTML = `
<style>
  #toolbar {
    background: rgba(255,255,255,0.2) url(https://app.posthog.com/static/images/a949629634c2b364328cc60c9a3eb3c2.svg) bottom right no-repeat;
    background-size: 100%;
		transform: scale(calc(1 / ${zoom}));
    transform-origin: top right;
    width: calc(${100 - 100 * zoom}vw - 40px / ${zoom});
		height: calc(100vh - 40px);
    top: 0;
    right: calc(-${100 - 100 * zoom}vw / ${zoom} + 40px / ${zoom});
    position: absolute;
    z-index:999999999;
    padding: 20px;
    box-sizing: border-box;
    font-size: 24px;
  }
</style>
<div id='toolbar'>
  Hello, I'm a toolbar!
</div>`;
  }
});
toolbar = document.createElement('posthog-toolbar');
document.body.appendChild(toolbar);
window.addEventListener('scroll', function(e) {
  toolbar.shadowRoot.getElementById('toolbar').style.marginTop = `${window.pageYOffset / zoom}px`;
})
```

It seems to work well on many of the sites I tested:

![image](https://user-images.githubusercontent.com/53387/82888413-8b7ee680-9f49-11ea-91ed-e34681c2cdce.png)
![image](https://user-images.githubusercontent.com/53387/82888429-92a5f480-9f49-11ea-9def-d64a05625467.png)
![image](https://user-images.githubusercontent.com/53387/82888450-9c2f5c80-9f49-11ea-8a31-2f76332d70ea.png)
![image](https://user-images.githubusercontent.com/53387/82888515-b2d5b380-9f49-11ea-8f61-6d4b627aec53.png)

The only issue is that we are zoomed out to 70% inside the <body>. This might or might not be what we want. We can increase the zoom inside the body tag with this:

```css
body > * {
  zoom: calc(1 / 0.7);
}
```

However that causes a few other elements to position themselves in a weird way (e.g. the floating top bar in google is way too large). I'm not sure if we need it as well, although it would be nice to see the text with the normal font size.

## Conclusion

In summary, what was envisioned on the mockups is possible, yet it will break pages slightly and zoom them out. Many of the unsolved problems from above can probably be addressed somehow. Plus if we add a "report layout issue" button that users can send feedback with, we can improve all or most broken elements over time.

However, there might or will still be cases where this type of toolbar will not work. For them we should offer a fallback to the only 100% guaranteed-to-work solution: the floating toolbar, like we have now.

Should we continue with this approach? Comments/questions/thoughts? :)

**Comment by timgl** [comment]
This is amazing and I think totally worth continuing
1. Agree that we should have a super easy fallback to a floating toolbar just in case.
2. On websites that don't specifically set a background (like bbc.co.uk) it should automatically set a white background
3. On my 2019 MBA scrolling is quite slow, which looking at the scroll function doesn't surprise me (as it's selecting all elements and doing something to them). Should be fairly easy to improve.

**Comment by mariusandra** [comment]
I edited the code above and added "white" as the background colo**u**r for all bodies that had none. It works now with BBC, although the "cookie banner" that they had went all the way to the end of the long page.

![image](https://user-images.githubusercontent.com/53387/82894897-a5bdc200-9f53-11ea-9613-07e96763d4a9.png)

Regarding the slow scrolling, unfortunately that's a reality we can't do much about... I think. You aren't supposed to manually add `marginTop` on scroll to elements... and should instead use `position:fixed` especially to avoid this slowness. Yet due to some DOM magic I need to work around `position:fixed`.

I can try using `transform: translate()`, which might speed things up a bit, yet from my past experience I know that it'll still be jittery. I'll need a transform-fallback anyway for elements that already have a marginTop set, so I'll try it.

Any alternative thoughts on how to solve this are more than welcome! :)

...

I forgot to add before, that one approach I also tried was to put the scaled down `body` in a box with `overflow: scroll` like to:

![image](https://user-images.githubusercontent.com/53387/82895188-182ea200-9f54-11ea-868a-6e8b8b0bdf14.png)

While I think this would solve some of the issues we had with scrolling and so on, unfortunately this won't work since the toolbar will also be inside body and you must scroll right to find it. I couldn't find any way to break out of a scroll container. All approaches I found revolved around adding a parent/child element and moving DOM nodes around.

Unfortunately you can't add a second body and we can't put the body in a div as that will break all `body > .something` CSS selectors which are used often enough to not go for this.

**Comment by timgl** [comment]
Closing this in favour of #868 and #870 as an epic ticket
